lib/Doctrine/ORM/Mapping/AssociationMapping.php

```diff
@@ -0,0 +1,240 @@
+     * (inheritance) parent <em>entity</em> class. The value is the FQCN of the
+     * topmost entity class that contains this association. (If there are
+     * transient classes in the class hierarchy, these are ignored, so the
+     * class property may in fact come from a class further up in the PHP class
+     * hierarchy.) To-many associations initially declared in mapped
+     * superclasses are <em>not</em> considered 'inherited' in the nearest
+     * entity subclasses.
+     *
+     * @var class-string
+     */
+    public string|null $inherited = null;
+
+    /**
+     * This is set when the association does not appear in the current class
+     * for the first time, but is initially declared in another parent
+     * <em>entity or mapped superclass</em>. The value is the FQCN of the
+     * topmost non-transient class that contains association information for
+     * this relationship.
+     */
+    public string|null $declared = null;
+
+    public array|null $cache = null;
+
+    public bool|null $id = null;
+
+    public bool $isCascadeRemove  = false;
+    public bool $isCascadePersist = false;
+    public bool $isCascadeRefresh = false;
+    public bool $isCascadeMerge   = false;
+    public bool $isCascadeDetach  = false;
+
+    public bool|null $isOnDeleteCascade = null;
+
+    /** @var array<string, string> */
+    public array|null $joinColumnFieldNames = null;
+
+    /** @var list<mixed> */
+    public array|null $joinTableColumns = null;
+
+    /** @var class-string */
+    public string|null $originalClass = null;
+
+    /** @var string */
+    public string|null $originalField = null;
+
+    public bool|null $orphanRemoval = null;
+
+    public bool|null $unique = null;
+
+    /**
+     * @param string       $fieldName    The name of the field in the entity
+     *                                   the association is mapped to.
+     * @param class-string $sourceEntity The class name of the source entity.
+     *                                   In the case of to-many associations
+     *                                   initially present in mapped
+     *                                   superclasses, the nearest
+     *                                   <em>entity</em> subclasses will be
+     *                                   considered the respective source
+     *                                   entities.
+     * @param class-string $targetEntity The class name of the target entity.
+     *                                   If it is fully-qualified it is used as
+     *                                   is. If it is a simple, unqualified
+     *                                   class name the namespace is assumed to
+     *                                   be the same as the namespace of the
+     *                                   source entity.
+     */
+    final public function __construct(
+        public string $fieldName,
+        public string $sourceEntity,
+        public string $targetEntity,
```

Comment:
Can we make those `readonly`? I don't think we want to mutate them.

Reply:
Not `$sourceEntity` according to the test suite, because of
https://github.com/doctrine/orm/blob/13d86957808f3fab3dfb9c681c59984a2bb4e4dd/lib/Doctrine/ORM/Mapping/ClassMetadataFactory.php#L444

It works for the other ones though.